.devcontainer/devcontainer.json

```diff
@@ -10,9 +10,6 @@
     },
     "ghcr.io/devcontainers/features/aws-cli:1": {}
   },
-  "postCreateCommand": [
-    "pip install --user -U pre-commit && pre-commit install-hooks",
-    "npm install -g aws-cdk@2.151.0"
-  ],
+  "postCreateCommand": "npm install -g aws-cdk@2.151.0 && pip install --user -U pre-commit && git config --global --add safe.directory \"$PWD\" && pre-commit install-hooks",
```

Comment:
could we not install the pip requirements here as well?  `pip install --upgrade pip && pip install -r requirements.txt -r requirements-dev.txt`?

Reply:
Done, see the updated changelog

Reply:
The linter run by the CI/CD workflow let me know about No newline at end of file. Could this be captured by the pre-commit checks?

Reply:
I don't understand your question.  The CI runs the exact same [pre-commit checks](https://github.com/Sage-Bionetworks-IT/openchallenges/blob/dev/.pre-commit-config.yaml) that the pre-commit hook runs.  It is capturing no newline at end of files.

Reply:
I was able to commit and push files that have no new line. The local linter didn't capture this issue; the linter executed by the GH workflow did. How is that possible?

Reply:
I don't know, but you can verify in your environment yourself.  Add a line to the README file with a new line at the end then run pre-commit on it.  like this..

```
(.venv) ➜  openchallenges git:(dev) ✗ pre-commit --version
pre-commit 3.8.0

(.venv) ➜  openchallenges git:(dev) ✗ vim README.md     # add a new line to the end of the file

(.venv) ➜  openchallenges git:(dev) ✗ pre-commit run --all-files
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing README.md

mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
flake8...................................................................Passed
yamllint.................................................................Passed
AWS CloudFormation Linter................................................Passed
black....................................................................Passed
Validate GitHub Workflows................................................Passed
Validate GitHub Actions..............................(no files to check)Skipped
```

Reply:
Do you run pre-commit manually? We could setup git hooks to run it automatically.

Reply:
No, `pre-commit install-hooks` which you have added in `setup.sh` sets up a commit hook which should run on every commit.  That's what i rely on to run the linters.  here's a test..

```
(.venv) ➜  openchallenges git:(test) ✗ pre-commit --version
pre-commit 3.8.0

(.venv) ➜  openchallenges git:(test) ✗ echo "test\n" >> README.md

(.venv) ➜  openchallenges git:(test) ✗ git commit -am "test commit hook"
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- exit code: 1
- files were modified by this hook

Fixing README.md

mixed line ending........................................................Passed
trim trailing whitespace.................................................Passed
flake8...............................................(no files to check)Skipped
yamllint.............................................(no files to check)Skipped
AWS CloudFormation Linter............................(no files to check)Skipped
black................................................(no files to check)Skipped
Validate GitHub Workflows............................(no files to check)Skipped
Validate GitHub Actions..............................(no files to check)Skipped
(.venv) ➜  openchallenges git:(test) ✗
```

Reply:
The hook does not work inside the dev container.

Committing a new file:

```
vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ source .venv/bin/activate
(.venv) vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ echo PLOP > PLOP
(.venv) vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ ls
app.py  BAR  cdk.json  docker  FOO  LICENSE.txt  openchallenges  PLOP  README.md  requirements-dev.txt  requirements.txt  source.bat  tests  tools
(.venv) vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ git add PLOP
(.venv) vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ git commit -m 'add PLOP'
[mark-repo-as-safe-directory 9491722] add PLOP
 1 file changed, 1 insertion(+)
 create mode 100644 PLOP
```

Committing an updated file:

```
(.venv) vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ echo "test\n" >> README.md
(.venv) vscode ➜ /workspaces/openchallenges-aws-cdk (mark-repo-as-safe-directory) $ git commit -am "test commit hook"
[mark-repo-as-safe-directory 7cb22b0] test commit hook
 2 files changed, 3 insertions(+)
```

Reply:
Running `pre-commit install` solves the issue.

Reply:
See https://github.com/pre-commit/pre-commit.com/issues/255#issuecomment-520494566.

I've updated the command to install the git hooks in `tools/setup.sh`.